### DISCUSSION TITLE
Docker datasource digest updates with release notes

### DISCUSSION THREAD

**Discussion by pyunramura** [renovatebot/renovate#16826] [Request Help]
I'm just wondering if there is any feasible way to pull in release notes for PRs that update docker image digests. Currently struggling quite a bit with finding a proper way to keep linuxserver packages up to date with the latest release as most of their tagging system is extremely out of spec. 

After combing through most of the discussions here of people struggling to regex their way around linuxserver's `:amd64-3.0.8.1678-r3-ls592-development` tags in their images I've decided to drop the proprietary tags and just stick to `:latest` with the digest pinned and let renovate keep it up to date. 

However I'd still like to see there is a way I can have the vcs_url in the image manifest referenced in any new PRs. Even if I wouldn't be able to directly reference the updated digest against the release notes, I'd still be able to compare them chronologically and pull in a diff if needed. 

Is this possible?

**Comment by PhilipAbed** [comment]
i dont understand exactly what you need, is this the same?
https://github.com/renovatebot/renovate/issues/16011

also do you know this?
https://docs.renovatebot.com/presets-docker/#dockerpindigests

**Reply by pyunramura** [reply]
I'm not sure if it #16011 is needed for this specific use case, but I do think it would be very useful to pull in change logs of packages with no vcs_url in their manifests. I apologize as my original post wasn't very clear on what exactly I'm trying to accomplish.

I currently have a set of docker swarm stack files that I'm using Renovate to keep updated. In lieu of a better solution at the time, I've been referencing the 'latest' tag and pulling the sha256 digest of that as I'm creating the stack file. Short example below.

`codeserver/docker-compose.yml`
```
---
version: '3.8'
services:
  codeserver:
    image: ghcr.io/linuxserver/code-server@sha256:8e9db910df43fc9c546d3005bad357f6c165c3e60260ce5d613763750bbf6e9d # This is the sha256 of the "latest" and "4.5.1-ls130" tags.
...
...
```

This setup has been working well for me for the last 6 months, but as I'm moving to the self-hosted variant of Renovate I had a few sore points I wanted to address in the process. Specifically  being able to check a named semver tag against a set of release notes that Renovate has pulled in when populating the PR. For example `image: ghcr.io/linuxserver/code-server:4.5.1` wherein Renovate will pull the image manifest, find the `org.opencontainers.image.source: "https://github.com/linuxserver/docker-code-server"`, and then populate the newly created PR with release notes pulled in from this url. This is the standard behavior for Renovate when used against packages that follow a proper semantic versioning scheme. 

Unfortunately linuxserver images just append the up-stream project's versioning scheme with their own '-ls' tags such as `code-server:4.5.1-ls130`; note that this is their "immutable" tagging scheme. This would work, with some 'versioning' regex, to extract the major.minor.patch.build version from the package, but this method breaks down with many of their other images such as `radarr:4.1.0.6175-ls147` , `paperless-ng:ng-1.5.0-ls46` , or `requestrr:V2.1.2-ls44`. Since I have around a dozen of these images from linuxserver and another dozen by various other authors that do not follow semver spec, I would strongly prefer to avoid having to set up individual package manager versioning schemes for each.

So after much testing and reading through the documentation and discussions here, I've given up more or less and decided to stay with using the digest of the 'latest' tag and have Renovate keep that up to date. The way I have had Renovate configured the last six months, when I get a new PR there are no release notes populated in the body of the PR. I'm not sure if that is a misconfiguration on my part or if that is by design/necessity, and I'm wondering if there is a way of configuring Renovate to pull in release notes for a PR whether it is using a named tag like `4.5.1` or a digest that is referenced by a mutable tag.

Sorry for the wordy reply, I thought it would be more helpful to give you more background on what I've been doing so you'd be able to know what exactly I'm trying to do.

And of course I am completely open to a better way of approaching this. I just want to keep the images up to date with the latest stable version, while keeping release notes, and preferably as immutable as possible.

As for `pinDigests`, am I correct in thinking that it will be useful for updating Dockerfiles specifically, or is there other uses for it outside that context?

**Reply by viceice** [reply]
the big issue is, there is jo API to get the tags pointing to a docker digest. that would need to make at least a head request for each docker tag in every renovate run. that's a no-go for renovate.

maybe we can in future somehow use the image metadata to return the version from image label, but there is currently no possible inside renovate to use that values.
that would require a big fat refactor and a pretty good design for multi datasource usage
